spotbugs/src/main/java/edu/umd/cs/findbugs/detect/CollectionIteratorCheck.java

```diff
@@ -0,0 +1,467 @@
+                    if (cycleVa != null && iteratorSetOrListName != null) {
+                        boolean flag = checkListOrSetValid(cycleVa, iteratorSetOrListName, locationList, i,
+                                classContext,
+                                method, constPool);
+                        if (!flag) {
+                            if (getClassOrMethodName(constPool, insIndex, true).endsWith("Set")) {
+                                fillBugReport(location, method, "set");
+
+                            } else if (getClassOrMethodName(constPool, insIndex, true).endsWith("List")) {
+                                fillBugReport(location, method, "list");
+                            }
+
+                        }
+                    }
+
+                }
+            }
+        }
+    }
+
+    private void fillBugReport(Location location, Method method, String type)
+            throws DataflowAnalysisException, CFGBuilderException {
+        if (null == location) {
+            return;
+        }
+
+        InstructionHandle insHandle = location.getHandle();
+        MethodGen methodGen = classCtx.getMethodGen(method);
+        String sourceFile = classCtx.getJavaClass().getSourceFileName();
+        ValueNumberDataflow valueNumDataFlow = classCtx.getValueNumberDataflow(method);
+
+        ValueNumberFrame vnaFrame = valueNumDataFlow.getFactAtLocation(location);
+        ValueNumber valueNumber = vnaFrame.getTopValue();
+
+        BugAnnotation variableAnnotation = ValueNumberSourceInfo.findAnnotationFromValueNumber(method, location,
+                valueNumber, vnaFrame, "VALUE_OF");
+
+        SourceLineAnnotation sourceLineAnnotation = SourceLineAnnotation.fromVisitedInstruction(classCtx, methodGen,
+                sourceFile, insHandle);
+        BugInstance bug;
+        if ("set".equals(type)) {
+            bug = new BugInstance(this, "SPEC_SET_REMOVEORADD_CHECK", NORMAL_PRIORITY);
+        } else if ("list".equals(type)) {
+            bug = new BugInstance(this, "SPEC_LIST_REMOVEORADD_CHECK", NORMAL_PRIORITY);
+        } else {
+            bug = new BugInstance(this, "SPEC_MAP_REMOVEORPUT_CHECK", NORMAL_PRIORITY);
+        }
+        bug.addClassAndMethod(methodGen, sourceFile);
+        bug.addOptionalAnnotation(variableAnnotation);
+        bug.addSourceLine(sourceLineAnnotation);
+        bugReporter.reportBug(bug);
+    }
+
+    private static boolean checkListOrSetValid(LocalVariable cycleVa, String iteratorSetOrListName,
+            List<Location> locationList,
+            int startIndex, ClassContext classContext, Method method, ConstantPoolGen constPool)
+            throws DataflowAnalysisException, CFGBuilderException {
+        boolean valid = true;
+        int startLoop = cycleVa.getStartPC();
+        int endLoop = cycleVa.getLength() + startLoop;
+        for (int i = startIndex; i < locationList.size(); i++) {
```

Comment:
This is a very long loop, and it contains multiple instances of `continue` and `break`, which is a code smell. Can it be refactored?